node/src/chain/stall_chain_fsm.rs

```diff
@@ -57,27 +70,41 @@ impl<DB: database::DB, N: Network, VM: VMExecution> StalledChainFSM<DB, N, VM> {
             acc,
         };
 
-        sm.update_tip(tip.inner().header().clone());
+        sm.update_tip(tip.inner().header());
         sm
     }
 
-    pub(crate) fn reset(&mut self, tip: Header) {
+    /// Attempts to reset the FSM state, if tip has changed
+    pub(crate) fn try_reset(&mut self, tip: &Header) {
         if self.tip.0.hash != tip.hash {
+            // Tip has changed, which means a new block is accepted either due
+            // to normal block acceptance or fallback execution
             self.update_tip(tip);
             self.state_transition(State::Running);
         }
     }
 
+    /// Handles heartbeat event
+    pub(crate) async fn on_heartbeat_event(&mut self) -> &State {
+        trace!(event = "chain.heartbeat",);
+        self.on_block_received(None).await
```

Comment:
I don't think we need to call `on_block_received`. Any tip update should have already triggered `on_block_received`, since it is a consequence of a block received (being it the tip+1 in insync, or the one we accepted after a revert).
I think we should simply call `on_running` here

Reply:
This would also make the use of `Option` in `on_block_received` unnecessary

Reply:
I can call 'on_running' , but the logic of tip-change detection will not be executed. 

Also you can call on_running if we are in `Running` state only.

Reply:
The point is that there's no need to call tip-change logic here, since the fms tip will be updated when actually accepting the new tip. So we can safely skip it here and just check how much time it passed since the last tip update. 
`on_running` does not check the state, so it will execute even if we're actually on Stalled.
You can rename it if you want to be coherent